virtual_rainforest/main.py

```diff
@@ -0,0 +1,126 @@
+"""Defines the function used to run a full simulation of the model.
+
+As well as setting up the function to run the overall virtual rainforest simulation,
+this script also defines the command line entry points for the model.
+"""
+
+from copy import deepcopy
+from typing import Any, Optional, Type, Union
+
+from virtual_rainforest.core.config import validate_config
+from virtual_rainforest.core.logger import LOGGER, log_and_raise
+from virtual_rainforest.core.model import MODEL_REGISTRY, BaseModel, InitialisationError
+
+
+def select_models(model_list: list[str]) -> Optional[list[Type[BaseModel]]]:
+    """Select the models to be run for a specific virtual rainforest simulation.
+
+    This function looks for models from a list of models, if these models can all be
+    found in the registry then they are returned. Otherwise an error is logged, which
+    should be handled appropriately downstream.
+
+    Args:
+        model_list: A list of models to select
+
+    Returns:
+        modules: A set of models to be configured
+    """
+
+    # Remove "core" from model list as it is not a model
+    if "core" in model_list:
+        model_list.remove("core")
+
+    LOGGER.info(f"Attempting to configure the following models: {model_list}")
```

Comment:
The optimal style for formatting logging messages seems to be to still use `%s` style formats - mostly because the error messages doesn't get built unless the message is emitted. Here, I think f-strings are always evaluated. The costs in this case (and many cases) are trivial but we should be consistent in what we use where. I prefer f-strings for general compactness and readability, but I've been using `%s` style for logging because of the deferred/non-evaluation.

https://docs.python.org/3/howto/logging.html#optimization

Thoughts?

Reply:
Does using f-strings prevent deferred evaluation? Nothing in that optimisation doc seems to suggest that they do, but I admitted don't have a firm grasp of how python strings work so might be missing something obvious

Reply:
Yep definitely see the value in avoiding expensive calls for logging output that won't actually ever be emitted. However, with their example:

```python
if logger.isEnabledFor(logging.DEBUG):
    logger.debug('Message with %s, %s', expensive_func1(),
                                        expensive_func2())
```

I was wondering if the optimisation was really to do with not using an f-string or rather nesting the logging call within a conditional statement? Like would the below not accomplish the same thing (particularly if the function calls had arguments)? Though admittedly in a less compact way

```python
if logger.isEnabledFor(logging.DEBUG):
    output1 = expensive_func1()
    output2 = expensive_func2()
    logger.debug(f"Message with {output1}, {output2}")
```

Reply:
That would work as:

```
if logger.isEnabledFor(logging.DEBUG):
    logger.debug(f"Message with {expensive_func1()}, {expensive_func2()}")
```

The optimisation link isn't particularly clear - this one is better: https://klichx.dev/2019/11/09/f-strings-and-python-logging/

Reply:
Ahh right so by not using f-strings you can avoid the evaluations when they are not gated behind an `if` statement, which is very useful. Definitely think we should switch then